### DISCUSSION TITLE
The flow rate of horizontal vs vertical gaps

### DISCUSSION THREAD

**Discussion by cshingle** [FakeFishGames/Barotrauma#8328] [Unstable feedback]
tl;dr: Horizontal gaps have a flow rate based on the difference on the difference in water levels between the two hulls. Vertical gaps have a static flow rate. The flow volume of a hatch is around 533.3334 which seems low when a standard pump has a flow rate of 600. With my specific setup water flows through horizontal gaps 11.1 times faster than it flows through vertically gaps of the same size. 

Anyone who has used the Sub Editor for any length of time has noticed that vertical and horizontal gaps have significantly different flow rates. I have actually started designing ships around this fact. Getting water to a ballast tank is much more difficult than removing it from the ballast tank due to the limited flow rate of hatches and ducts. 

To explore this I broke out the editor and debugger. I created two identical sized hulls. One with a hatch on the bottom and one with a door in the bottom right corner scaled to 0.310 which gives it an identical gap height as the hatches gap width (8 grid lines or 128 units). I then created a drain hull for each with a pump set to a pump rate of 10,000 to ensure that no water builds up. I created a circuit that fills both hulls with water and then opens the door/hatch and measures how long they take to drain. I also verified a bunch of things via the debugger.

First here are the results.
Hull A drained in 2 seconds (Horizontal)
Hull B drained in 14 seconds. (Vertical)

1) The game gets the gap size by reading the gap height when vertical or width when horizontal. (Pretty obvious)
2) A door and a hatch of the same size open at identical rates. (Yes the flow rate is affected by how far open the door/hatch is during that tick.)
3) The flow rate of horizontal gaps (doors) is calculated by the difference in available water volume in each hull. 
This is the formula used by the game. There is more to it then this, but this outlines my point.
https://github.com/Regalis11/Barotrauma/blob/master/Barotrauma/BarotraumaShared/SharedSource/Map/Gap.cs#L398
`delta = Math.Min((hull1.Pressure - (hull2.Pressure + subOffset.Y)) * 5.0f * sizeModifier, Math.Min(hull1.WaterVolume, hull1.Volume));`
delta = 5935.58144

These are the real numbers in my test
sizeModifier = 1.28 (This is due to the gap being 128 units wide/tall)
hull1.Pressure = 14943.4346
hull2.Pressure = 14016
(hull1.Pressure - (hull2.Pressure + subOffset.Y)) = 927.4346
927.4346 * 5 * 1.28 = 5935.58144

4) The flow rate of vertical gaps (hatches) is calculated only by the size of the gap.
https://github.com/Regalis11/Barotrauma/blob/master/Barotrauma/BarotraumaShared/SharedSource/Map/Gap.cs#L468
`float delta = Math.Min(hull1.WaterVolume, deltaTime * 25000f * sizeModifier);`
delta = 533.3334

sizeModifier = 1.28
deltaTime = 0.016666669
25000 is a magic number that controls flow rate that goes back to the beginning of the gap.cs file.
5) For completeness the max flow rate for water flowing from a lower hull through a hatch to a hull above it is 170.66669056
https://github.com/Regalis11/Barotrauma/blob/master/Barotrauma/BarotraumaShared/SharedSource/Map/Gap.cs#L438
float delta = Math.Min(hull2.WaterVolume - hull2.Volume + (hull2.Volume * Hull.MaxCompress), deltaTime * 8000.0f * sizeModifier);

sizeModifier = 1.28
deltaTime = 0.016666669
8000 is a magic number.
6) These values are only for hull to hull flow only. There is different code for flow from no hull into a hull.

With my specific setup water flows horizontally 11.1 times faster than it flows vertically. I expect that the larger the hulls are the more the flow imbalance is. It feels ridiculous to me that an open hatch can't even move as much water as a default standard pump (533.3334 vs 600). A duct is able to somewhat outpace a default small pump 133.333335 vs 100.
I expect that these slow flow rates where chosen at some point during the games development to make subs flood more slowly. I personally feel like these values should be higher, but don't know how gameplay would be affected. On that note I suggest that these two magic numbers or some multiplier be moved to XML so moders can mess with these values so we can see how it affects gameplay.
